numpy/core/fromnumeric.py

```diff
@@ -1276,6 +1276,104 @@ def argmin(a, axis=None, out=None):
     return _wrapfunc(a, 'argmin', axis=axis, out=out)
 
 
+def _topk_dispatcher(a, k, axis=None, largest=None, sorted=None):
+    return (a,)
+
+
+@array_function_dispatch(_topk_dispatcher)
+def topk(a, k, axis=-1, largest=True, sorted=True):
+    """
+    Finds values and indices of the `k` largest/smallest 
+    elements along the given `axis`.
+
+    Parameters
+    ----------
+    a: array_like
+        Array with given axis at least k.
+    k: int
+        Number of top elements to look for along the given axis.
+    axis: int or None, optional
+        Axis along which to find topk. If None, the array is flattened 
+        before sorting. The default is -1 (the last axis).
+    largest: bool, optional
+        Controls whether to return largest or smallest elements. 
+    sorted: bool, optional
+        If true the resulting k elements will be sorted by the values.
+
+    Returns
+    -------
+    topk_values : ndarray
+        Array of values of `k` largest/smallest elements 
+        along the specified `axis`.
+    topk_indices: ndarray, int
+        Array of indices of `k` largest/smallest elements 
+        along the specified `axis`.
+
+    See Also
+    --------
+    sort : Describes sorting algorithms used.
+    argsort : Indirect sort.
+    partition : Describes partition algorithms used.
+    argpartition : Indirect partial sort.
+    take_along_axis : Take values from the input array by 
+                      matching 1d index and data slices.
+
+    Examples
+    --------
+    One dimensional array:
+
+    >>> x = np.array([3, 1, 2])
+    >>> np.topk(x)
+    (array([3, 2]), array([0, 2], dtype=int64))
+
+    Two-dimensional array:
+
+    >>> x = np.array([[0, 3, 4], [2, 2, 1], [5, 1, 2]])
+    >>> val, ind = np.topk(x, 2, axis=1) # along the last axis
+    >>> val
+    [[4 3]
+     [2 2]
+     [5 2]]
+    >>> ind
+    [[2 1]
+     [0 1]
+     [0 2]]
+
+    >>> val, ind = np.topk(x, 2, axis=None) # along the flattened array
+    >>> val)
+    [5 4]
+    >>> ind
+    [6 2]
+
+    >>> val, ind = np.topk(x, 2, axis=0) # along the first axis
+    >>> val
+    [[5 3 4]
+     [2 2 2]]
+    >>> ind
+    [[2 0 0]
+     [1 1 2]]
+    """
+    if largest:
+        index_array = np.argpartition(-a, k-1, axis=axis, order=None)
+    else:
+        index_array = np.argpartition(a, k-1, axis=axis, order=None)
```

Comment:
Negating like this won't work for unsigned types

Reply:
Thanks for your reply, I will test it for unsigned types again.

Reply:
Assume that `x` and `y` are both unsigned int type, `-x` (or `-y`) is acually equal to `~x + 1 = 2^n - x` (or `~y + 1 = 2^n - y`), `n` is the bit size of the unsigned int type. You can test it with the following code:
```python
x = np.random.randint(1, 100, size=(32, 64), dtype=np.uint32)
minus_x_v1 = -x
minus_x_v2 = ~x + 1
minus_x_v3 = 2**32 - x
print(minus_x_v1.dtype, minus_x_v2.dtype, minus_x_v3.dtype)
print(np.allclose(minus_x_v1, minus_x_v2))
print(np.allclose(minus_x_v1, minus_x_v3))
    
```
If `x > y`, then `2^n - x < 2^n - y` and `~x + 1 < ~y + 1`, i.e., `-x < -y`.